### PR TITLE
Only give remote queues 1 minute to flush samples on shutdown.

### LINKED ISSUE #2972: Remote storage blocks shutdown if other end is having issues
From @brian-brazil in #2847:

> The remote write generally doesn't seem to like the other end having issues, such as being down for 5+ minutes. In this scenario, it also blocks shutdown.

I propose we have a timeout (1 min) to allow samples to flush, and otherwise exit uncleanly.  WDYT>

### LINKED ISSUE #3707: Prometheus 2.0 reload configuration issues
**What did you do?**
Prometheus sends samples to our remote storage. One time when remote storage is going to down, we see in logs:
level=warn ts=2018-01-19T08:29:08.516631616Z caller=queue_manager.go:485 component=remote msg="Error sending samples to remote storage" count=1 err="Post <storage_url>: getsockopt: no route to host"
If we try to reload prometheus configuration by sending post on <prometheus_url>/-/reload  when remote storage is down, this request is never finished. Looks like prometheus trying to send data to storage again and again, before reload configuration.
**What did you expect to see?**
Prometheus should reload configuration
**What did you see instead? Under which circumstances?**
Configuration wasn't reloaded, no response on POST <prometheus_url>/-/reload request 
**Environment**

* System information:

	Linux 3.10.0-514.26.2.el7.x86_64 x86_64

* Prometheus version:

	2.0.0

* Prometheus configuration file:
```
global:
  scrape_interval: 15s
  scrape_timeout: 10s
  evaluation_interval: 15s
  external_labels:
    monitor: default
...
remote_write:
- url: <remote_storage_url>
  remote_timeout: 30s
  write_relabel_configs:
  - source_labels: [indicatorName]
    separator: ;
    regex: (.+)
    replacement: $1
    action: keep
  queue_config:
    capacity: 100000
    max_shards: 1000
    max_samples_per_send: 100
    batch_send_deadline: 5s
    max_retries: 10
    min_backoff: 30ms
    max_backoff: 100ms
```

* Logs:
```
2018/01/19 08:25:22 Redirected: /-/reload
level=info ts=2018-01-19T08:25:22.313752379Z caller=main.go:490 msg="Loading configuration file" filename=/config/prometheus/default.yml
level=info ts=2018-01-19T08:25:22.319062591Z caller=queue_manager.go:253 component=remote msg="Stopping remote storage..."
level=warn ts=2018-01-19T08:25:23.406480826Z caller=queue_manager.go:485 component=remote msg="Error sending samples to remote storage" count=8 err="Post <remote_storage_url>: getsockopt: no route to host"
level=warn ts=2018-01-19T08:25:26.412528632Z caller=queue_manager.go:485 component=remote msg="Error sending samples to remote storage" count=8 err="Post <remote_storage_url>: getsockopt: no route to host"
level=warn ts=2018-01-19T08:25:29.418470232Z caller=queue_manager.go:485 component=remote msg="Error sending samples to remote storage" count=8 err="Post <remote_storage_url>: getsockopt: no route to host"
```

### PR DESCRIPTION
**Pull request by tomwilkie**
Fixes #2972, fixes #3707

~This is just a POC - needs to be made flag controlled and need to refactor out the waitgroup, as the extra goroutine is a bit ugly.~

### CODE REVIEW

**Comment by tomwilkie** [comment]
(CircleCI failures are unrelated)

File: config/config.go [hunk 1]
```diff
@@ -690,6 +692,10 @@ type QueueConfig struct {
 	// On recoverable errors, backoff exponentially.
 	MinBackoff time.Duration `yaml:"min_backoff,omitempty"`
 	MaxBackoff time.Duration `yaml:"max_backoff,omitempty"`
+
+	// On shutdown or config reload allow the following duration for flushing
+	// pending samples, otherwise continue without waiting.
+	FlushDeadline time.Duration `yaml:"flush_deadline"`
```
Comment (brian-brazil) [inline config/config.go]: I don't think this should be in the config file, it's more a flag thing.

**Comment by krasi-georgiev** [comment]
do we even need to make the timeout settable?

LGTM and I don't have an opinion for config vs flag so it is up to @brian-brazil to give green light.

**Comment by brian-brazil** [comment]
If it is settable, it must be a flag. No point in waiting less time for one source when another has a longer timeout, ans this is a very flagg-y setting.

**Comment by krasi-georgiev** [comment]
@tomwilkie shall we leave this hard coded and if someone requests to be settable to revisit?

**Comment by tomwilkie** [comment]
(Closing in favour of #4187)

**Pull request closed without merging**